### ISSUE TITLE
[Feature][AzureDevOps/GitHub] Support extracting Azure Pipelines metrics from GitHub Repositories

### ISSUE THREAD

**Issue by Feggah** [apache/devlake#5692]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar feature requirement.

### Use case

I have GitHub Repositories (Data Scopes of GitHub Plugin) and I use Azure Pipelines as a CI/CD tool. Azure Pipelines is part of AzureDevOps.

I want to be able to select GitHub Repositories and configure DevLake to extract DORA metrics from Azure Pipelines + GitHub Repo. But, today, I can only use GitHub Actions with GitHub Repositories.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by Startrekzky** [comment]
Got it. May I know why you use this combination of GitHub repo and Azure Pipeline? Are you maintaining an open-source project? I want to know if it is a common use case.

**Comment by Feggah** [comment]
We (as a company) decided to use GitHub as the tool to version our repositories. As we grew bigger, each team/tribe decided to use a different CI/CD tool to manage their applications.

This means that we have all possible combinations of tools, for example:
- Github Repos + GitHub Actions (supported by DevLake)
- GitHub Repos + Azure Pipelines (not supported, that's my feature request)
- GitHub Repos + CircleCI
- etc...

I don't know if this is a common use case in another companies (I imagine it is), but here we have 2k+ applications with this usecase.

**Comment by Startrekzky** [comment]
Thank you, @Feggah . This is pretty informative. I just found out we seem to have supported this feature in https://github.com/apache/incubator-devlake/issues/5249. However, I didn't find any docs about how to configure it. 

@CamilleTeruel Do you have any idea how can we use this feature?

**Comment by keon94** [comment]
@Feggah We do support Github-exclusive repos. They are prefixed with `GitHub` in the data-scopes seletion window. Example:
![image](https://github.com/apache/incubator-devlake/assets/25063936/c0e1367d-b93e-4be3-91f3-7d2fd671c91a)

**Comment by Feggah** [comment]
I cannot reproduce your test @keon94 . Is there any special configuration?

I have a AzureDevOps connection, but I can only see Azure Repositories. Do I need to select something special?

**Comment by keon94** [comment]
@Feggah There is no special config. Is this a private repo?
Try this for troubleshooting (which is what the backend code does). Use POSTMAN or curl using your Azure devops token to:
1. Call `GET https://dev.azure.com/${YOUR_ORG_NAME}/${YOUR_PROJECT_NAME}/_apis/serviceendpoint/endpoints`. Look in the returned JSON for the field `type`, and make sure you see `GitHub`, and the field `id` for it.
2. Call `GET https://dev.azure.com/${YOUR_ORG_NAME}/${YOUR_PROJECT_NAME}/_apis/sourceProviders/GitHub/repositories?serviceEndpointId=${ID_FROM_ABOVE}` and see if your github repo is in there.

**Comment by Feggah** [comment]
> Is this a private repo?

Yes, they are private

---

Using the POSTMAN it worked. I got a lot of repositories from AzureDevOps API. See below:

GET https://dev.azure.com/REDACTED/REDACTED/_apis/sourceProviders/GitHub/repositories?serviceEndpointId=REDACTED
![image](https://github.com/apache/incubator-devlake/assets/11353912/f163be5c-d81f-41fb-a453-119817a5b6b9)

But I also have a lot of ServiceConnections (or ServiceEndpoints) in my AzureDevOps Project. There are multiple GitHub connections. Maybe that's why it is not displaying in DevLake?

https://dev.azure.com/REDACTED/REDACTED/_apis/serviceendpoint/endpoints
![image](https://github.com/apache/incubator-devlake/assets/11353912/eac418ab-b2b5-4b61-8c2a-1007c6c6811b)

Some thoughts:
- Maybe DevLake is selecting a ServiceConnection without the necessary permissions?
- The GET request that returned 2000 repositories took 40.2 seconds. Maybe DevLake is timeouting the request?

**Comment by Startrekzky** [comment]
Hi @keon94 , I tested it in v0.18.0-alpha17 (around beta1 I guess?) and the results were:

1. When adding a data scope to an Azure connection, I can fetch the external Github repo
![image](https://github.com/apache/incubator-devlake/assets/14050754/80643c0d-20e6-4c8e-9b9e-f09a726b7e05)

2. However, after I added it to the connection, it didn't show the repo's name. 
![image](https://github.com/apache/incubator-devlake/assets/14050754/265fd72a-d106-4549-b404-45602bf3f81b)

3. When I tried to delete the scope, it was not deletable.
![20230726200321_rec_](https://github.com/apache/incubator-devlake/assets/14050754/458ec3f0-fb9a-45cf-9a2e-098eed9d19c8)

4. When I tried to add the connection to a project and collect the data of the external repo, it threw an error:
```
error patching the blueprint (400) Wraps: (2) Wraps: (3) make plan for blueprint failed (400) Wraps: (4) Invalid scope id (400)"
```
which seemed to have something to do with the scope id.
![image](https://github.com/apache/incubator-devlake/assets/14050754/36ea3972-7702-41f7-8800-97c885ee6985)

**Comment by Startrekzky** [comment]
@keon94 Got it. I'll re-test it in beta3 and let you know the results.

**Comment by Feggah** [comment]
@Startrekzky could you also check if you can see this GitHub repository when creating a Blueprint instead of adding directly to the Connection?

Because I can't see any GitHub repos from an AzureDevOps Connection when creating a blueprint. Maybe there is a bug in that flow

**Comment by Startrekzky** [comment]
Hi @Feggah , the v0.18.0's flow of adding repos has changed. I can only add data scope (external repos in this case) in the Data Connection page now.

And I just verified that in v0.18.0-beta3, when I tried to add repos to a Azure DevOps connection, it can list out the external GitHub repos(see the pic). cc. @keon94 
![image](https://github.com/apache/incubator-devlake/assets/14050754/3a1760e2-cdbb-4eb8-a95d-55beb5c63b1e)

@Feggah Can you show me the screenshot of adding the external data scope in your v0.17.0-beta13? Perhaps it's a bug in that version. I'm not sure.

**Comment by Feggah** [comment]
@Startrekzky I upgraded to v0.18.0-beta3 and tried to add this external repository. The problem now is that the repositories are loading forever:
![image](https://github.com/apache/incubator-devlake/assets/11353912/2491b842-0a5f-42c6-b740-dcb32ba67fd1)

By looking at the logs of the DevLake API:
![image](https://github.com/apache/incubator-devlake/assets/11353912/98ca395a-cccf-4131-b6f6-2fbc2b1304b5)

I think the problem is that my AzureDevOps project has a lot of Service Connections with different types. The fix probably would be skipping Service Connections which are not type == GitHub? Then it will only use valid Service Connections to find repositories

**Comment by Startrekzky** [comment]
Hi @Feggah, sorry that I missed the notifications. I'll forward it to @klesh to take a look.

**Comment by klesh** [comment]
@CamilleTeruel Can you take a look at this issue?

**Comment by CamilleTeruel** [comment]
fixed via  #5950

**Comment by Startrekzky** [comment]
Hi @Feggah , this issue is fixed in https://github.com/apache/incubator-devlake/releases/tag/v0.18.0-beta8 . Please let me know if you still have this problem.
